tests/rwqueue.cpp

```diff
@@ -0,0 +1,144 @@
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "rwqueue.h"
+
+#include <gtest/gtest.h>
+
+#include <thread>
+
+namespace {
+
+constexpr auto iterations = 10000;
+
+TEST(RWQueue, EnqueueDequeue)
+{
+	RWQueue<int> q(65);
+	for (int iteration = 0; iteration != 128;
+	     ++iteration) { // check there's no problem with mismatch
+		            // between nominal and allocated capacity
+		EXPECT_EQ(q.MaxCapacity(), 65);
+		EXPECT_EQ(q.Size(), 0);
+		EXPECT_TRUE(q.IsEmpty());
+		q.CopyIn(0);
+		EXPECT_EQ(q.MaxCapacity(), 65);
+		EXPECT_EQ(q.Size(), 1);
+		EXPECT_FALSE(q.IsEmpty());
+		for (int i = 1; i != 65; ++i)
+			q.MoveIn(i);
+		EXPECT_EQ(q.Size(), 65);
+		EXPECT_FALSE(q.IsEmpty());
+
+		// Basic dequeue
+		int item;
+		item = q.PopOut();
+		EXPECT_EQ(item, 0);
+		for (int i = 1; i != 65; ++i) {
+			item = q.PopOut();
+			EXPECT_EQ(item, i);
+		}
+		EXPECT_EQ(item, 64);
+		EXPECT_TRUE(q.IsEmpty());
+	}
+}
+
+TEST(RWQueue, ZeroCapacity)
+{
+	// Zero capacity
+	EXPECT_DEBUG_DEATH({ RWQueue<int> q(0); }, "");
+}
+
+void rw_consume(RWQueue<int> *q, const size_t *max_depth, bool *got_mismatch)
+{
+	int item;
+	for (int i = 0; i != iterations; ++i) {
+		EXPECT_TRUE(q->Size() <= *max_depth);
+		item = q->PopOut();
+		if (item != i && !*got_mismatch) {
+			*got_mismatch = true;
+		}
+	}
+}
+
+void rw_produce_copy(RWQueue<int> *q, const size_t *max_depth)
+{
+	for (int i = 0; i != iterations; ++i) {
+		q->CopyIn(i);
+		EXPECT_TRUE(q->Size() <= *max_depth);
+	}
+}
+
+void rw_produce_move(RWQueue<int> *q, const size_t *max_depth)
+{
+	for (int i = 0; i != iterations; ++i) {
+		q->MoveIn(i);
+		EXPECT_TRUE(q->Size() <= *max_depth);
+	}
+}
+
+TEST(RWQueue, ThreadedCopyIn)
+{
+	const size_t max_depth = 8;
+	bool got_mismatch = false;
+
+	RWQueue<int> q(max_depth);
+
+	std::thread writer(rw_produce_copy, &q, &max_depth);
+	std::thread reader(rw_consume, &q, &max_depth, &got_mismatch);
+
+	writer.join();
+	reader.join();
+
+	// Make sure we've consumed all produced items and the queue is empty
+	EXPECT_EQ(q.Size(), 0);
+
+	// Make sure there wasn't a single out-of-sequence item consumed
+	EXPECT_FALSE(got_mismatch);
+
+	// Confirm both threads report as terminated
+	EXPECT_FALSE(reader.joinable());
+	EXPECT_FALSE(writer.joinable());
+}
+
+TEST(RWQueue, ThreadedMoveIn)
+{
+	const size_t max_depth = 8;
+	bool got_mismatch = false;
+
+	RWQueue<int> q(max_depth);
+
+	std::thread writer(rw_produce_move, &q, &max_depth);
+	std::thread reader(rw_consume, &q, &max_depth, &got_mismatch);
+
+	writer.join();
+	reader.join();
+
+	// Make sure we've consumed all produced items and the queue is empty
+	EXPECT_EQ(q.Size(), 0);
+
+	// Make sure there wasn't a single out-of-sequence item consumed
+	EXPECT_FALSE(got_mismatch);
+
+	// Confirm both threads report as terminated
+	EXPECT_FALSE(reader.joinable());
+	EXPECT_FALSE(writer.joinable());
```

Comment:
This comment is misleading. `joinable()` does not report if thread is terminated. It reports if thread is, well, joinable.

Both threads were joined already, so they are surely not joinable any more - you are testing if STL behaves according to specification in these two EXPECT lines.

*Applies to previous test as well.*

Reply:
> you are testing if STL behaves according to specification in these two EXPECT lines.

Back when I added the moodycamel queue, gtest sometimes reported that these threads were still present when testing completed (despite being joined), so I added these join-checks just to prove to myself that they were functionally joined.

I retested today though, and I'm not able to reproduce this "thread-is-still-present" issue; so dropping these checks!